src/premiumvideodevice/PremiumVideoStreamHandler.ts

```diff
@@ -0,0 +1,211 @@
+import Logger from '../logger/Logger';
+import PremiumVideoEffectDriver from './PremiumVideoEffectDriver';
+
+const DEFAULT_FRAMERATE = 15;
+
+/** @internal */
+interface HTMLCanvasElementWithCaptureStream extends HTMLCanvasElement {
+  // Not in IE, but that's OK.
+  captureStream(frameRate?: number): MediaStream;
+}
+
+/**
+ * [[PremiumVideoStreamHandler]] Will be used to handle the stream associated with the
+ * device in the DefaultVideoTransformDevice. It is responsible for triggering the
+ * action in the PremiumVideoEffectDriver and processing the transformed frames onto
+ * the output canvas
+ */
+export default class PremiumVideoStreamHandler {
+  private framerate: number = DEFAULT_FRAMERATE;
+  private lastTimeOut: ReturnType<typeof setTimeout> | undefined;
+
+  // Inputs
+  private videoInput: HTMLVideoElement = document.createElement('video') as HTMLVideoElement;
+  private canvasInput: HTMLCanvasElement = document.createElement('canvas');
+  private inputCtx = this.canvasInput.getContext('2d');
+  private inputVideoStream: MediaStream | null = null;
+
+  // Outputs
+  private canvasOutput: HTMLCanvasElementWithCaptureStream = document.createElement(
+    'canvas'
+  ) as HTMLCanvasElementWithCaptureStream;
+  private outputCtx = this.canvasOutput.getContext('2d');
+  outputMediaStream: MediaStream = new MediaStream();
+
+  // Constructor just sets a logger for the object
+  constructor(private logger: Logger, private videoEffectDriver: PremiumVideoEffectDriver) {}
+
+  /**
+   * If existing, return the output media stream. If not existing, create and return
+   * a media stream off of our output canvas
+   * @returns MediaStream
+   */
+  getActiveOutputMediaStream(): MediaStream {
+    if (this.isOutputMediaStreamActive()) {
+      return this.outputMediaStream;
+    }
+    this.outputMediaStream = this.canvasOutput.captureStream(this.framerate);
+    this.cloneInputAudioTracksToOutput();
+    return this.outputMediaStream;
+  }
+
+  /**
+   * Configure an inputMediaStream so that it is placed onto a HTML video element and
+   * configure our transformation function (apply) to process as our stream continues
+   * loading new data
+   */
+  async setInputMediaStream(inputMediaStream: MediaStream | null): Promise<void> {
+    if (!inputMediaStream) {
+      this.stop();
+      return;
+    }
+
+    if (inputMediaStream.getVideoTracks().length === 0) {
+      this.logger.error('No video tracks in input media stream, ignoring');
+      return;
+    }
+
+    this.inputVideoStream = inputMediaStream;
```

Comment:
I feel like we can group line 68->82 as a new function

Reply:
Added new private async function startVideoOnMediaStream()